libs/constrained-generators/src/Constrained/ListSplit.hs

```diff
@@ -0,0 +1,164 @@
+prop_merge_sound =
+  withMaxSuccess 10000 $
+  forAll arbitrary $ \ (pf1 :: ListSplit Int) ->
+  forAll arbitrary $ \ pf2 ->
+  case merge pf1 pf2 of
+    Left _ -> discard
+    Right pf3 -> forAll (genFromListSplit pf3) $ \ x ->
+        counterexample (unlines [show pf1, show pf2, show pf3, show x]) $
+          classify (nullListSplit pf1) "null pf1" $
+          classify (nullListSplit pf2) "null pf2" $
+          classify (nullListSplit pf3) "null pf3" $
+          conformSplit x pf1 && conformSplit x pf2 && conformSplit x pf3
+
+nullListSplit :: ListSplit a -> Bool
+nullListSplit (Split p s) = null $ fromMaybe [] p ++ fromMaybe [] s
+
+genSmall :: Arbitrary a => Gen (Maybe [a])
+genSmall =
+  frequency
+    [
+      ( 3
+      , Just <$> do
+          n <- choose (1, 5)
+          vectorOf n (resize 5 arbitrary)
+      )
+    , (1, pure Nothing)
+    ]
+
+instance Arbitrary a => Arbitrary (ListSplit a) where
+  arbitrary = Split <$> genSmall <*> genSmall
+
+-- | if (merge x y) == Right z, then anything that conforms to z, should also conform to both x and y
+merge :: (Eq a, Show a) => ListSplit a -> ListSplit a -> Either (NE.NonEmpty String) (ListSplit a)
+merge (Split Nothing Nothing) x = Right x
+merge x (Split Nothing Nothing) = Right x
+merge (Split (Just ps) Nothing) (Split (Just xs) Nothing)
+  | isPrefixOf ps xs = Right (Split (Just xs) Nothing)
+  | isPrefixOf xs ps = Right (Split (Just ps) Nothing)
+  | True = Left $ NE.fromList ["Non matching prefixes in merge" ++ show ps ++ " " ++ show xs]
+merge (Split (Just ps) Nothing) (Split Nothing (Just ys)) = Right (mergeOverlap ps ys)
+merge p@(Split (Just ps) Nothing) q@(Split (Just xs) (Just ys))
+  | isPrefixOf ps (xs ++ ys) = Right q
+  | True = Left $ NE.fromList ["Non matching ListSplit in merge", "  " ++ show p, " " ++ show q]
+merge (Split Nothing (Just ss)) (Split (Just xs) Nothing) = Right (mergeOverlap xs ss)
+merge (Split Nothing (Just ss)) (Split Nothing (Just ys))
+  | isSuffixOf ss ys = Right (Split Nothing (Just ys))
+  | isSuffixOf ys ss = Right (Split Nothing (Just ss))
+  | True = Left $ NE.fromList ["Non matching suffixes in merge" ++ show ss ++ " " ++ show ys]
+merge p@(Split Nothing (Just ss)) q@(Split (Just xs) (Just ys))
+  | isSuffixOf ss (xs ++ ys) = Right q
+  | True = Left $ NE.fromList ["Non matching ListSplit in merge", "  " ++ show p, " " ++ show q]
+merge p@(Split (Just ps) (Just ss)) (Split q@(Just xs) Nothing)
+  | isPrefixOf xs (ps ++ ss) = Right p
+  | True = Left $ NE.fromList ["Non ListSplit in merge", "  " ++ show p, " " ++ show q]
+merge p@(Split (Just ps) (Just ss)) q@(Split Nothing (Just ys))
+  | isSuffixOf ys (ps ++ ss) = Right p
+  | True = Left $ NE.fromList ["Non ListSplit in merge", "  " ++ show p, " " ++ show q]
+merge p@(Split (Just ps) (Just ss)) q@(Split (Just xs) (Just ys)) =
+  if (ps ++ ss) == (xs ++ ys)
+    then Right p
+    else Left $ NE.fromList ["Two rigid ListSplits are inconsistent.", "  " ++ show p, "  " ++ show q]
```

Comment:
For consistency with all other cases:
```suggestion
  | (ps ++ ss) == (xs ++ ys) = Right p
  | otherwise = Left $ NE.fromList ["Two rigid ListSplits are inconsistent.", "  " ++ show p, "  " ++ show q]
```